### DISCUSSION TITLE
Custom Screenshot Writer doesn't work correctly

### DISCUSSION THREAD

**Discussion by serhatozdursun** [getgauge/gauge#2152] [General]
java -version
openjdk version "17.0.1" 2021-10-19

Gauge version: 1.4.1
Plugins
-------
html-report (4.1.2)
java (0.7.15)
screenshot (0.1.0)
spectacle (0.1.4)
xml-report (0.2.3)

I have overridden takeScreenshot() methods as below

public class CustomGaugeScreenShot implements CustomScreenshotWriter {

    @Override
    public String takeScreenshot() {

        var driver = PlatformManager.getInstances().getDriver();
        File scrFile = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
        String fileName = "/" + new RandomHelper().hexCodeOfRandomness(32) + ".png";
        var gaugeScreenShotDir = System.getenv("gauge_screenshots_dir");
        File destFile = new File(gaugeScreenShotDir + fileName);
        try {
            FileUtils.copyFile(scrFile, destFile);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return destFile.getAbsolutePath();
    }
}

Screenshots taken by Appium are saved under its ".gauge/screenshots".

![image](https://user-images.githubusercontent.com/12559494/140606132-42a0724b-5469-4c44-bd0c-279d5e067fc6.png)

But screenshot src in the report like this:

![image](https://user-images.githubusercontent.com/12559494/140606196-312f30c2-3656-4810-8ae7-01dd5cc9db6a.png)

in the nutshell, gauge reports seek the screenshots in the images folder below the html-reports folder with the path returned by takeScreenshot() method.

**Comment by serhatozdursun** [comment]
I changed my code like below and it's work but I think it's not difficult like that

` @Override
    public String takeScreenshot() {

        var driver = PlatformManager.getInstances().getDriver();
        File scrFile = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
        String fileName = new RandomHelper().hexCodeOfRandomness(32) + ".png";
        var gaugeScreenShotDir = System.getenv("gauge_reports_dir") + "/html-report/images/";
        File destFile = new File(gaugeScreenShotDir + fileName);
        try {
            FileUtils.copyFile(scrFile, destFile);
        } catch (IOException e) {
            e.printStackTrace();
        }
        var name = destFile.getName().split("/");
        return name[name.length - 1];
    }`

**Comment by zabil** [comment]
Have you tried using the recommended code which is

```
public class CustomScreenshotGrabber implements CustomScreenshotWriter {

    // Return a screenshot file name
    @Override
    public String takeScreenshot() {
        TakesScreenshot driver = (TakesScreenshot) DriverFactory.getDriver();
        String screenshotFileName = String.format("screenshot-%s.png", UUID.randomUUID().toString());
        try {
            Files.write(Path.of(System.getenv("gauge_screenshots_dir"), screenshotFileName),
                    driver.getScreenshotAs(OutputType.BYTES));
        } catch (IOException e) {
            e.printStackTrace();
        }
        return screenshotFileName;
    }

}
```
Notice that it uses `Files.write(Path.of(...));` which is a better way to build the correct path. Gauge simply reads whatever path is returned from this function.

**Reply by serhatozdursun** [reply]
Thanks for your advice. I found the solution before however, the recommended code is better than mine. I will use it.
